### ISSUE TITLE
Add `AllChem.MMFFGetMoleculeForceField().CalcGradient()` to Python wrappers

### ISSUE THREAD

**Issue by theavey** [rdkit/rdkit#1688]
**Description:**
You can currently get the force field energy and do a minimization through the Python interface, but there is no direct way to access the gradient/force information.

I would like to use rdkit to do MD with MMFF, but that is more difficult without the gradients being available to Python. I've seen some preliminary code from @ptosco on interfacing rdkit with OpenMM and implementing MMFF in OpenMM, but neither is in stable releases of those codes.

- RDKit Version: 2017.09.2
- Platform: Linux (CentOS)

```python
mol = Chem.MolFromPDBFile('3htmf.pdb')
pyMP = AllChem.MMFFGetMoleculeProperties(mol)
pyFF = AllChem.MMFFGetMoleculeForceField(mol, pyMP)
print(pyFF.CalcEnergy())
grad = pyFF.CalcGradient()   # This is not a defined method.
# or
grad = np.zeroes([mol.GetNumAtoms(), 3])
pyFF.CalcGradient(grad)
```

It would be great to be able to call this method in both ways it's declared in the C++ code: by just passing in an array to be filled or by passing in an array to be filled and the atomic positions.
Returning the gradients as a list or array would work just as well (but might be a bit slower).

**Comment by ptosco** [comment]
@theavey  Hi Thomas, you are correct that the OpenMM code is not yet in production stage, as I haven't found the time to polish it up yet,
I should be able to expose the `CalcGradient` method to the Python layer later tonight and submit a PR. I will let you know once it is done.

**Comment by kunkinger** [comment]
Another related issue on gradients of MMFF94. I modified testMMFFForceField.cpp to the following code so that it just calculate the energy and gradient before and after minimization. 

```
using namespace RDKit;
int main( ) {
  std::string molBlock =
      "butane\n"
      "     RDKit          3D\n"
      "butane\n"
      " 17 16  0  0  0  0  0  0  0  0999 V2000\n"
      "    0.0000    0.0000    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0\n"
      "    1.4280    0.0000    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0\n"
      "    1.7913   -0.2660    0.9927 H   0  0  0  0  0  0  0  0  0  0  0  0\n"
      "    1.9040    1.3004   -0.3485 C   0  0  0  0  0  0  0  0  0  0  0  0\n"
      "    1.5407    2.0271    0.3782 H   0  0  0  0  0  0  0  0  0  0  0  0\n"
      "    1.5407    1.5664   -1.3411 H   0  0  0  0  0  0  0  0  0  0  0  0\n"
      "    3.3320    1.3004   -0.3485 C   0  0  0  0  0  0  0  0  0  0  0  0\n"
      "    3.6953    1.5162   -1.3532 H   0  0  0  0  0  0  0  0  0  0  0  0\n"
      "    3.8080    0.0192    0.0649 C   0  0  0  0  0  0  0  0  0  0  0  0\n"
      "    3.4447   -0.7431   -0.6243 H   0  0  0  0  0  0  0  0  0  0  0  0\n"
      "    3.4447   -0.1966    1.0697 H   0  0  0  0  0  0  0  0  0  0  0  0\n"
      "    4.8980    0.0192    0.0649 H   0  0  0  0  0  0  0  0  0  0  0  0\n"
      "    3.6954    2.0627    0.3408 H   0  0  0  0  0  0  0  0  0  0  0  0\n"
      "    1.7913   -0.7267   -0.7267 H   0  0  0  0  0  0  0  0  0  0  0  0\n"
      "   -0.3633    0.7267    0.7267 H   0  0  0  0  0  0  0  0  0  0  0  0\n"
      "   -0.3633   -0.9926    0.2660 H   0  0  0  0  0  0  0  0  0  0  0  0\n"
      "   -0.3633    0.2660   -0.9926 H   0  0  0  0  0  0  0  0  0  0  0  0\n"
      "  1  2  1  0  0  0  0\n"
      "  1 15  1  0  0  0  0\n"
      "  1 16  1  0  0  0  0\n"
      "  1 17  1  0  0  0  0\n"
      "  2  3  1  0  0  0  0\n"
      "  2  4  1  0  0  0  0\n"
      "  2 14  1  0  0  0  0\n"
      "  4  5  1  0  0  0  0\n"
      "  4  6  1  0  0  0  0\n"
      "  4  7  1  0  0  0  0\n"
      "  7  8  1  0  0  0  0\n"
      "  7  9  1  0  0  0  0\n"
      "  7 13  1  0  0  0  0\n"
      "  9 10  1  0  0  0  0\n"
      "  9 11  1  0  0  0  0\n"
      "  9 12  1  0  0  0  0\n"
      "M  END\n";

  {
    RDKit::RWMol *mol = RDKit::MolBlockToMol(molBlock, true, false);
    RDKit::RWMol *cmol = new RDKit::RWMol(*mol);

    MMFF::MMFFMolProperties *mmffMolProperties =
        new MMFF::MMFFMolProperties(*mol);
    ForceFields::ForceField *field =
        RDKit::MMFF::constructForceField(*mol, mmffMolProperties);
    TEST_ASSERT(field);
    field->initialize();
    std::cout << "Number of Atom: " << field->positions().size() << std::endl;
    std::cout << "Begin of MMFF94 Charge"<< std::endl;
    for (unsigned int i = 0; i < field->positions().size(); ++i) {
	std::cout<<mmffMolProperties->getMMFFPartialCharge(i)<<std::endl;
    }
    std::cout << "End of MMFF94 Charge"<<std::endl;
    std::cout << "MMFF94 Energy: " << field->calcEnergy() << std::endl;
    double *grad = new double[3*field->positions().size()];
    field->calcGrad(grad);
    std::cout.precision(15);
    std::cout <<"Begin of MMFF94 Grads"<<std::endl;
    for (unsigned int i = 0; i < field->positions().size(); ++i) {
      std::cout<<grad[3*i]<<"  "<<grad[3*i+1]<<"  "<<grad[3*i+2]<<std::endl;
    }
    std::cout <<"End of MMFF94 Grads"<<std::endl;
    field->minimize();
    std::cout << "After minize Energy :" << field->calcEnergy() << std::endl;
   
    double *grad2 = new double[3*field->positions().size()];
    field->calcGrad(grad2);
    std::cout <<"Printing Optimized Grads"<<std::endl;
    for (unsigned int i = 0; i < field->positions().size(); ++i) {
      std::cout<<grad2[3*i]<<"  "<<grad2[3*i+1]<<"  "<<grad2[3*i+2]<<std::endl;
    }
    std::cout <<"End of Optimized Grads"<<std::endl;
  }
  return 0;
}
```
Gradients after optimization are still several kcal/mol/Angstrom ( Here I assume the unit of gradient is kcal/mol/Angstrom). Same issue if I use an external optimizer and other molecules. 
I am wondering if it is just the gradients of MMFF94 that I get from rdkit are noisy or I am doing something wrong? Thanks.

**Comment by kunkinger** [comment]
I also find the Numerical Gradient and Analytical Gradient of MMFF94 do not match. They have difference around few kcal/mol/Angstrom. The numerical gradient seems to be the right one since it can optimize molecule to the right minimal energy geometry.


Below are the two gradient of aspirin. (RDKit Version: 2017.09.2, Platform: ubuntu)
```
Begin of MMFF94 Numerical Grads
-0.041768481651161  -0.00750544586480828  0.0740559585210576
0.0166266545420513  0.0746433613585396  0.00345067974194535
0.086820551814526  0.0256774285389838  0.059189972034801
0.0273776663561875  0.0277530375569768  0.0407298671234457
-0.043869039245692  0.00874438228493091  0.0799230349457503
0.0633454760645691  -0.0377020626274316  -0.0137003780764644
-0.0113106125354534  0.0338480163009081  -0.0418346530750568
0.0446410012955312  -0.02326056005586  0.0546642809240438
0.00727022211322037  0.0401764793878101  -0.000122258825285826
-0.0126414818168996  -0.0685661802890536  -0.00807559104032407
-0.133696771342784  0.0198441810539407  -0.00821424208652388
-0.0335130884820956  0.0166757892827718  -0.0703225719433931
0.0441844122178736  -0.0153876303699008  -0.00217876880981294
0.0102052766237648  0.0170508709373962  -0.00159448774184057
0.00324517586136608  -0.00271943640939298  0.016743848973988
0.00285840009439653  0.0339565826124044  -0.0111753333698061
0.01478730915494  0.0298689045052924  -0.0369147201695341
0.0252263305355882  -0.0520621824762202  0.00443461658505839
-0.025438458806093  0.013597059478343  -0.0035927371300204
-0.014095041933615  -0.0238572511079838  -0.0427447901074629
0.0648156724736282  -0.0280683185138741  -0.0381836652962875
End of MMFF94 Numerical Grads
MMFF94 Energy: 18.9146633049144
Begin of MMFF94 Grads
-0.0471771794015237  -0.0086115241671223  0.0704541870785737
0.0119689685500189  -2.77061257351997  -0.446476581514837
1.15661720035728  -2.03368178587075  0.436569925347171
1.81870654917861  0.817777817722055  -1.45145198736759
-0.153024346118052  0.645661386345492  0.479507944458581
-0.716753971306034  -0.57658458403379  0.018959847234894
-0.789974026808805  1.89729359777745  0.416528940251551
-2.09400787566948  -0.477750711335088  -0.254256614617617
-2.12040327269158  1.97013195908018  0.116113913647854
-2.81927939752633  0.706337295440391  -0.277990294851398
-0.209091850095897  -1.82538462384432  0.00978878549652515
2.06567815328026  0.684767724054223  -0.386880768360533
3.55446238750981  0.527482844606501  0.180759450206229
-0.233768937906486  2.78417646391068  0.759844835527772
-2.68481093564605  -1.34021344762693  -0.58803802011042
-2.64243264533676  2.92090768862239  0.138641763053588
-3.8406418522648  0.863666619957444  -0.575393876857701
3.77517058272033  1.33212519875812  0.852721529319295
4.17366758737578  0.586589006052416  -0.669138558134927
3.62980261805417  -0.429348000896163  0.667008705815536
-0.180328042764163  -3.6527123656927  -0.459374963968618
End of MMFF94 Grads
```

**Comment by ptosco** [comment]
@kunkinger You need to initialise to zero the `grad` and `grad2` arrays before calling `calcGrad()`. Then you will obtain the expected ~0 gradient after minimisation.
May I please ask you next time to create a separate thread rather than hijacking an existing one? Keeping discussion threads and issues separate helps in relating issues to PRs more easily. Thanks!

**Comment by greglandrum** [comment]
Closed by #1694
